src/module.c

```diff
@@ -6544,11 +6547,27 @@ RedisModuleCallReply *RM_Call(RedisModuleCtx *ctx, const char *cmdname, const ch
      *
      * If RM_SetContextUser has set a user, that user is used, otherwise
      * use the attached client's user. If there is no attached client user and no manually
-     * set user, an error will be returned */
+     * set user, an error will be returned.
+     * An internal command should only succeed for an internal connection. */
     if (flags & REDISMODULE_ARGV_RUN_AS_USER) {
         int acl_errpos;
         int acl_retval;
 
+        // Internal commands succeed if their corresponding connection is internal as well.
+        if (c->cmd->flags & CMD_INTERNAL) {
+            if ((ctx->client->flags & CLIENT_INTERNAL) || mustObeyClient(ctx->client)) {
+                // Internal connections can execute internal commands, without
+                // ACL checks.
+                goto acl_ok;
+            } else {
+                // Non-internal connections cannot execute internal commands.
+                sds msg = sdscatfmt(sdsempty(), "unknown command '%.128s'", c->cmd);
```

Comment:
No need to free `msg`?

Reply:
No, it is free'd automatically later (registered via `autoMemoryAdd()` in `cleanup`).